{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "727be743-7a89-4242-83d3-3920884a7327",
   "metadata": {},
   "source": [
    "## Download the Sample Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b939766-26a3-4d9f-93f3-3f6062c42627",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The tar file is 9.3GB and, when extracted, an additional 14GB\n",
    "!mkdir sample_data\n",
    "%cd sample_data\n",
    "!wget -O parabricks_sample.tar.gz \"https://s3.amazonaws.com/parabricks.sample/parabricks_sample.tar.gz\"\n",
    "!tar xvf parabricks_sample.tar.gz\n",
    "!mv parabricks_sample/* .\n",
    "%cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9de4abb-0139-4022-87f5-eb90479a25cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a4076948-d6f0-4270-9418-378e115601c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data  Ref\n"
     ]
    }
   ],
   "source": [
    "!ls sample_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "af9bccea-d790-4073-9937-3c32718f1faf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample_1.fq.gz\tsample_2.fq.gz\n"
     ]
    }
   ],
   "source": [
    "!ls sample_data/Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4449d1d7-09a9-459d-b71d-7ccc75037583",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo_sapiens_assembly38.dict\n",
      "Homo_sapiens_assembly38.fasta\n",
      "Homo_sapiens_assembly38.fasta.amb\n",
      "Homo_sapiens_assembly38.fasta.ann\n",
      "Homo_sapiens_assembly38.fasta.bwt\n",
      "Homo_sapiens_assembly38.fasta.fai\n",
      "Homo_sapiens_assembly38.fasta.pac\n",
      "Homo_sapiens_assembly38.fasta.sa\n",
      "Homo_sapiens_assembly38.known_indels.vcf.gz\n",
      "Homo_sapiens_assembly38.known_indels.vcf.gz.tbi\n"
     ]
    }
   ],
   "source": [
    "!ls sample_data/Ref"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81d854c7-c48b-4e62-a843-eadcca5ab261",
   "metadata": {},
   "source": [
    "## GPU Monitoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d5523e52-751a-4059-9751-f36883738c2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Nov 29 07:06:26 2023       \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |\n",
      "|-----------------------------------------+----------------------+----------------------+\n",
      "| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |\n",
      "| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |\n",
      "|                                         |                      |               MIG M. |\n",
      "|=========================================+======================+======================|\n",
      "|   0  NVIDIA GeForce RTX 4090        Off | 00000000:07:00.0 Off |                  Off |\n",
      "| 33%   34C    P8              15W / 450W |      2MiB / 24564MiB |      0%      Default |\n",
      "|                                         |                      |                  N/A |\n",
      "+-----------------------------------------+----------------------+----------------------+\n",
      "                                                                                         \n",
      "+---------------------------------------------------------------------------------------+\n",
      "| Processes:                                                                            |\n",
      "|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |\n",
      "|        ID   ID                                                             Usage      |\n",
      "|=======================================================================================|\n",
      "|  No running processes found                                                           |\n",
      "+---------------------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "!nvidia-smi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0374156-867b-4477-a3bf-1bfcd17fef31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the command below in the terminal\n",
    "### watch -n 0.5 nvidia-smi\n",
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0832aacf-e020-411e-a3ae-73bae3d20553",
   "metadata": {},
   "source": [
    "## Alignment: FASTQ to BAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "59143bbc-005d-46d2-a9c0-3c27d03b7a25",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "usage: pbrun fq2bam [<args>]\n",
      "Help: pbrun fq2bam -h\n",
      "\n",
      "Run GPU-bwa mem, co-ordinate sorting, marking duplicates, and Base Quality\n",
      "Score Recalibration to convert FASTQ to BAM/CRAM.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input Output file options:\n",
      "  Options for Input and Output files for this tool:\n",
      "\n",
      "  --ref REF             Path to the reference file. (default: None)\n",
      "  --in-fq [IN_FQ [IN_FQ ...]]\n",
      "                        Path to the pair-ended FASTQ files followed by\n",
      "                        optional read groups with quotes (Example:\n",
      "                        \"@RG\\tID:foo\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:foo\").\n",
      "                        The files must be in fastq or fastq.gz format. All\n",
      "                        sets of inputs should have a read group; otherwise,\n",
      "                        none should have a read group, and it will be\n",
      "                        automatically added by the pipeline. This option can\n",
      "                        be repeated multiple times. Example 1: --in-fq\n",
      "                        sampleX_1_1.fastq.gz sampleX_1_2.fastq.gz --in-fq\n",
      "                        sampleX_2_1.fastq.gz sampleX_2_2.fastq.gz. Example 2:\n",
      "                        --in-fq sampleX_1_1.fastq.gz sampleX_1_2.fastq.gz\n",
      "                        \"@RG\\tID:foo\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:unit1\"\n",
      "                        --in-fq sampleX_2_1.fastq.gz sampleX_2_2.fastq.gz\n",
      "                        \"@RG\\tID:foo2\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:unit2\".\n",
      "                        For the same sample, Read Groups should have the same\n",
      "                        sample name (SM) and a different ID and PU. (default:\n",
      "                        None)\n",
      "  --in-se-fq [IN_SE_FQ [IN_SE_FQ ...]]\n",
      "                        Path to the single-ended FASTQ file followed by\n",
      "                        optional read group with quotes (Example:\n",
      "                        \"@RG\\tID:foo\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:foo\").\n",
      "                        The file must be in fastq or fastq.gz format. Either\n",
      "                        all sets of inputs have a read group, or none should\n",
      "                        have one, and it will be automatically added by the\n",
      "                        pipeline. This option can be repeated multiple times.\n",
      "                        Example 1: --in-se-fq sampleX_1.fastq.gz --in-se-fq\n",
      "                        sampleX_2.fastq.gz . Example 2: --in-se-fq\n",
      "                        sampleX_1.fastq.gz\n",
      "                        \"@RG\\tID:foo\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:unit1\"\n",
      "                        --in-se-fq sampleX_2.fastq.gz\n",
      "                        \"@RG\\tID:foo2\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:unit2\" .\n",
      "                        For the same sample, Read Groups should have the same\n",
      "                        sample name (SM) and a different ID and PU (default:\n",
      "                        None)\n",
      "  --in-fq-list IN_FQ_LIST\n",
      "                        Path to a file that contains the locations of pair-\n",
      "                        ended FASTQ files. Each line must contain the location\n",
      "                        of two FASTQ files followed by a read group, each\n",
      "                        separated by a space. Each set of files (and\n",
      "                        associated read group) must be on a separate line.\n",
      "                        Files must be in fastq/fastq.gz format. Line syntax:\n",
      "                        <fastq_1> <fastq_2> <read group> (default: None)\n",
      "  --knownSites KNOWNSITES\n",
      "                        Path to a known indels file. The file must be in\n",
      "                        vcf.gz format. This option can be used multiple times.\n",
      "                        (default: None)\n",
      "  --interval-file INTERVAL_FILE\n",
      "                        Path to an interval file in one of these formats:\n",
      "                        Picard-style (.interval_list or .picard), GATK-style\n",
      "                        (.list or .intervals), or BED file (.bed). This option\n",
      "                        can be used multiple times. (default: None)\n",
      "  --out-recal-file OUT_RECAL_FILE\n",
      "                        Path of a report file after Base Quality Score\n",
      "                        Recalibration. (default: None)\n",
      "  --out-bam OUT_BAM     Path of a BAM/CRAM file after Marking Duplicates.\n",
      "                        (default: None)\n",
      "  --out-duplicate-metrics OUT_DUPLICATE_METRICS\n",
      "                        Path of duplicate metrics file after Marking\n",
      "                        Duplicates. (default: None)\n",
      "  --out-qc-metrics-dir OUT_QC_METRICS_DIR\n",
      "                        Path of the directory where QC metrics will be\n",
      "                        generated. (default: None)\n",
      "\n",
      "Tool Options:\n",
      "  Options specific to the tool\n",
      "\n",
      "  -L INTERVAL, --interval INTERVAL\n",
      "                        Interval within which to call bqsr from the input\n",
      "                        reads. All intervals will have a padding of 100 to get\n",
      "                        read records, and overlapping intervals will be\n",
      "                        combined. Interval files should be passed using the\n",
      "                        --interval-file option. This option can be used\n",
      "                        multiple times (e.g. \"-L chr1 -L chr2:10000 -L\n",
      "                        chr3:20000+ -L chr4:10000-20000\"). (default: None)\n",
      "  --bwa-options BWA_OPTIONS\n",
      "                        Pass supported bwa mem options as one string. The\n",
      "                        current original bwa mem supported options are -M, -Y\n",
      "                        and -T (e.g. --bwa-options=\"-M -Y\") (default: None)\n",
      "  --no-warnings         Suppress warning messages about system thread and\n",
      "                        memory usage. (default: None)\n",
      "  --gpuwrite            Use one GPU to accelerate writing final BAM. (default:\n",
      "                        None)\n",
      "  --gpusort             Use GPUs to accelerate sorting and marking. (default:\n",
      "                        None)\n",
      "  --low-memory          Use low memory mode (default: None)\n",
      "  --filter-flag FILTER_FLAG\n",
      "                        Dont generate SAM entries in the output if the entry's\n",
      "                        flag's meet this criteria. Criteria: (flag & filter !=\n",
      "                        0) (default: 0)\n",
      "  --skip-multiple-hits  Filter SAM entries whose length of SA is not 0\n",
      "                        (default: None)\n",
      "  --min-read-length MIN_READ_LENGTH\n",
      "                        Skip reads below minimum read length. They will not be\n",
      "                        part of output (default: None)\n",
      "  --align-only          Generate output BAM after bwa-mem. The output will not\n",
      "                        be co-ordinate sorted or duplicates will not be marked\n",
      "                        (default: None)\n",
      "  --no-markdups         Do not perform the Mark Duplicates step. Return BAM\n",
      "                        after sorting. (default: None)\n",
      "  --fix-mate            Add mate cigar (MC) and mate quality (MQ) tags to the\n",
      "                        output file. (default: None)\n",
      "  --markdups-assume-sortorder-queryname\n",
      "                        Assume the reads are sorted by queryname for Marking\n",
      "                        Duplicates. This will mark secondary, supplementary,\n",
      "                        and unmapped reads as duplicates as well. This flag\n",
      "                        will not impact variant calling while increasing\n",
      "                        processing times. (default: None)\n",
      "  --markdups-picard-version-2182\n",
      "                        Assume marking duplicates to be similar to Picard\n",
      "                        version 2.18.2. (default: None)\n",
      "  --optical-duplicate-pixel-distance OPTICAL_DUPLICATE_PIXEL_DISTANCE\n",
      "                        The maximum offset between two duplicate clusters in\n",
      "                        order to consider them optical duplicates. Ignored if\n",
      "                        --out-duplicate-metrics is not passed. (default: None)\n",
      "  --read-group-sm READ_GROUP_SM\n",
      "                        SM tag for read groups in this run. (default: None)\n",
      "  --read-group-lb READ_GROUP_LB\n",
      "                        LB tag for read groups in this run. (default: None)\n",
      "  --read-group-pl READ_GROUP_PL\n",
      "                        PL tag for read groups in this run. (default: None)\n",
      "  --read-group-id-prefix READ_GROUP_ID_PREFIX\n",
      "                        Prefix for the ID and PU tags for read groups in this\n",
      "                        run. This prefix will be used for all pairs of fastq\n",
      "                        files in this run. The ID and PU tags will consist of\n",
      "                        this prefix and an identifier, that will be unique for\n",
      "                        a pair of fastq files. (default: None)\n",
      "  -ip INTERVAL_PADDING, --interval-padding INTERVAL_PADDING\n",
      "                        Amount of padding (in base pairs) to add to each\n",
      "                        interval you are including. (default: None)\n",
      "\n",
      "Run Options:\n",
      "  Options related to resource/verbosity/etc\n",
      "\n",
      "  --logfile LOGFILE     Path to the log file. If not specified, messages will\n",
      "                        only be written to the standard error output.\n",
      "                        (default: None)\n",
      "  --tmp-dir TMP_DIR     Full path to the directory where temporary files will\n",
      "                        be stored. (default: .)\n",
      "  --with-petagene-dir WITH_PETAGENE_DIR\n",
      "                        Full path to the PetaGene installation directory. By\n",
      "                        default, this should have been installed at\n",
      "                        /opt/petagene. Use of this option also requires that\n",
      "                        the PetaLink library has been preloaded by setting the\n",
      "                        LD_PRELOAD environment variable. Optionally set the\n",
      "                        PETASUITE_REFPATH and PGCLOUD_CREDPATH environment\n",
      "                        variables that are used for data and credentials.\n",
      "                        Optionally set the PetaLinkMode environment variable\n",
      "                        that is used to further configure PetaLink, notably\n",
      "                        setting it to \"+write\" to enable outputting compressed\n",
      "                        BAM and .fastq files. (default: None)\n",
      "  --keep-tmp            Do not delete the directory storing temporary files\n",
      "                        after completion. (default: None)\n",
      "  --no-seccomp-override\n",
      "                        Do not override seccomp options for docker. (default:\n",
      "                        None)\n",
      "  --version             View compatible software versions. (default: None)\n",
      "  --num-gpus NUM_GPUS   Number of GPUs to use for a run. (default: 1)\n"
     ]
    }
   ],
   "source": [
    "!pbrun fq2bam -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8a661081-f1c6-4e63-bfc5-e15eee8f2a31",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "\n",
      "\n",
      "[Parabricks Options Mesg]: Checking argument compatibility\n",
      "[Parabricks Options Mesg]: Automatically generating ID prefix\n",
      "[Parabricks Options Mesg]: Read group created for /workdir/sample_data/Data/sample_1.fq.gz and\n",
      "/workdir/sample_data/Data/sample_2.fq.gz\n",
      "[Parabricks Options Mesg]: @RG\\tID:HK3TJBCX2.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:HK3TJBCX2.1\n",
      "[PB Info 2023-Nov-29 07:01:32] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:01:32] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-29 07:01:32] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-29 07:01:32] ||                       GPU-BWA mem, Sorting Phase-I                       ||\n",
      "[PB Info 2023-Nov-29 07:01:32] ------------------------------------------------------------------------------\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[PB Info 2023-Nov-29 07:01:37] GPU-BWA mem\n",
      "[PB Info 2023-Nov-29 07:01:37] ProgressMeter\tReads\t\tBase Pairs Aligned\n",
      "[PB Info 2023-Nov-29 07:01:59] 5043564\t\t580000000\n",
      "[PB Info 2023-Nov-29 07:02:10] 10087128\t1160000000\n",
      "[PB Info 2023-Nov-29 07:02:20] 15130692\t1740000000\n",
      "[PB Info 2023-Nov-29 07:02:30] 20174256\t2320000000\n",
      "[PB Info 2023-Nov-29 07:02:39] 25217820\t2900000000\n",
      "[PB Info 2023-Nov-29 07:02:48] 30261384\t3480000000\n",
      "[PB Info 2023-Nov-29 07:02:56] 35304948\t4060000000\n",
      "[PB Info 2023-Nov-29 07:03:06] 40348512\t4640000000\n",
      "[PB Info 2023-Nov-29 07:03:14] 45392076\t5220000000\n",
      "[PB Info 2023-Nov-29 07:03:22] 50435640\t5800000000\n",
      "[PB Info 2023-Nov-29 07:04:13] \n",
      "GPU-BWA Mem time: 155.522691 seconds\n",
      "[PB Info 2023-Nov-29 07:04:13] GPU-BWA Mem is finished.\n",
      "\n",
      "[main] CMD: /usr/local/parabricks/binaries//bin/bwa mem -Z ./pbOpts.txt -F 0 /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta /workdir/sample_data/Data/sample_1.fq.gz /workdir/sample_data/Data/sample_2.fq.gz @RG\\tID:HK3TJBCX2.1\\tLB:lib1\\tPL:bar\\tSM:sample\\tPU:HK3TJBCX2.1\n",
      "[main] Real time: 160.898 sec; CPU: 1744.693 sec\n",
      "[PB Info 2023-Nov-29 07:04:13] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:04:13] ||        Program:                      GPU-BWA mem, Sorting Phase-I        ||\n",
      "[PB Info 2023-Nov-29 07:04:13] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-29 07:04:13] ||        Start Time:                       Wed Nov 29 07:01:32 2023        ||\n",
      "[PB Info 2023-Nov-29 07:04:13] ||        End Time:                         Wed Nov 29 07:04:13 2023        ||\n",
      "[PB Info 2023-Nov-29 07:04:13] ||        Total Time:                           2 minutes 41 seconds        ||\n",
      "[PB Info 2023-Nov-29 07:04:13] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:04:16] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:04:16] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-29 07:04:16] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-29 07:04:16] ||                             Sorting Phase-II                             ||\n",
      "[PB Info 2023-Nov-29 07:04:16] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:04:16] progressMeter - Percentage\n",
      "[PB Info 2023-Nov-29 07:04:16] 0.0\t 0.00 GB\n",
      "[PB Info 2023-Nov-29 07:04:21] 28.4\t 3.00 GB\n",
      "[PB Info 2023-Nov-29 07:04:26] 57.9\t 2.00 GB\n",
      "[PB Info 2023-Nov-29 07:04:31] 85.1\t 0.00 GB\n",
      "[PB Info 2023-Nov-29 07:04:36] Sorting and Marking: 20.002 seconds\n",
      "[PB Info 2023-Nov-29 07:04:36] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:04:36] ||        Program:                                  Sorting Phase-II        ||\n",
      "[PB Info 2023-Nov-29 07:04:36] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-29 07:04:36] ||        Start Time:                       Wed Nov 29 07:04:16 2023        ||\n",
      "[PB Info 2023-Nov-29 07:04:36] ||        End Time:                         Wed Nov 29 07:04:36 2023        ||\n",
      "[PB Info 2023-Nov-29 07:04:36] ||        Total Time:                                     20 seconds        ||\n",
      "[PB Info 2023-Nov-29 07:04:36] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:04:36] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:04:36] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-29 07:04:36] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-29 07:04:36] ||                         Marking Duplicates, BQSR                         ||\n",
      "[PB Info 2023-Nov-29 07:04:36] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:04:36] Using PBBinBamFile for BAM writing\n",
      "[PB Info 2023-Nov-29 07:04:36] progressMeter -\tPercentage\n",
      "[PB Info 2023-Nov-29 07:04:46] 34.0\t 10.60 GB\n",
      "[PB Info 2023-Nov-29 07:04:56] 75.4\t 4.48 GB\n",
      "[PB Info 2023-Nov-29 07:05:06] 100.0\t 0.00 GB\n",
      "[PB Info 2023-Nov-29 07:05:53] BQSR and writing final BAM:  76.960 seconds\n",
      "[PB Info 2023-Nov-29 07:05:53] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:05:53] ||        Program:                          Marking Duplicates, BQSR        ||\n",
      "[PB Info 2023-Nov-29 07:05:53] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-29 07:05:53] ||        Start Time:                       Wed Nov 29 07:04:36 2023        ||\n",
      "[PB Info 2023-Nov-29 07:05:53] ||        End Time:                         Wed Nov 29 07:05:53 2023        ||\n",
      "[PB Info 2023-Nov-29 07:05:53] ||        Total Time:                            1 minute 17 seconds        ||\n",
      "[PB Info 2023-Nov-29 07:05:53] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun fq2bam \\\n",
    "      --ref /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "      --in-fq /workdir/sample_data/Data/sample_1.fq.gz /workdir/sample_data/Data/sample_2.fq.gz \\\n",
    "      --out-bam /workdir/outputdir/fq2bam_output.bam\n",
    "      --num-gpus 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "481db6af-77bd-4398-b82a-d26b61c693a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fq2bam_output.bam  fq2bam_output.bam.bai  fq2bam_output_chrs.txt\n"
     ]
    }
   ],
   "source": [
    "!ls outputdir/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b640e24c-121d-417b-b7f6-c41f4b649d6a",
   "metadata": {},
   "source": [
    "## Variant Calling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2e1e51d-b363-4cff-8cfa-2dec27843353",
   "metadata": {},
   "source": [
    "#### GATK Haplotypecaller"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4c4fb290-443b-403a-a72c-964912ca0f4c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "usage: pbrun haplotypecaller <options>\n",
      "Help: pbrun haplotypecaller -h\n",
      "\n",
      "Run HaplotypeCaller to convert BAM/CRAM to VCF.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input Output file options:\n",
      "  Options for Input and Output files for this tool:\n",
      "\n",
      "  --ref REF             Path to the reference file. (default: None)\n",
      "  --in-bam IN_BAM       Path to the input BAM/CRAM file for variant calling.\n",
      "                        The argument may also be a local folder containing\n",
      "                        several bams; each will be processed by 1 GPU in batch\n",
      "                        mode. (default: None)\n",
      "  --in-recal-file IN_RECAL_FILE\n",
      "                        Path to the input BQSR report. (default: None)\n",
      "  --interval-file INTERVAL_FILE\n",
      "                        Path to an interval file in one of these formats:\n",
      "                        Picard-style (.interval_list or .picard), GATK-style\n",
      "                        (.list or .intervals), or BED file (.bed). This option\n",
      "                        can be used multiple times. (default: None)\n",
      "  --out-variants OUT_VARIANTS\n",
      "                        Path of the vcf/g.vcf/gvcf.gz file after variant\n",
      "                        calling. The argument may also be a local folder in\n",
      "                        batch mode. (default: None)\n",
      "\n",
      "Tool Options:\n",
      "  Options specific to the tool\n",
      "\n",
      "  --haplotypecaller-options HAPLOTYPECALLER_OPTIONS\n",
      "                        Pass supported haplotype caller options as one string.\n",
      "                        The following are currently supported original\n",
      "                        haplotypecaller options: -A\n",
      "                        <AS_BaseQualityRankSumTest, AS_FisherStrand,\n",
      "                        AS_InbreedingCoeff, AS_MappingQualityRankSumTest,\n",
      "                        AS_QualByDepth, AS_RMSMappingQuality,\n",
      "                        AS_ReadPosRankSumTest, AS_StrandOddsRatio,\n",
      "                        BaseQualityRankSumTest, ChromosomeCounts,\n",
      "                        ClippingRankSumTest, Coverage, DepthPerAlleleBySample,\n",
      "                        DepthPerSampleHC, ExcessHet, FisherStrand,\n",
      "                        InbreedingCoeff, MappingQualityRankSumTest,\n",
      "                        QualByDepth, RMSMappingQuality, ReadPosRankSumTest,\n",
      "                        ReferenceBases, StrandBiasBySample, StrandOddsRatio,\n",
      "                        TandemRepeat>,-AX <same options as -A>,--output-mode\n",
      "                        <EMIT_VARIANTS_ONLY, EMIT_ALL_CONFIDENT_SITES,\n",
      "                        EMIT_ALL_ACTIVE_SITES> ,-max-reads-per-alignment-start\n",
      "                        <int>, -min-dangling-branch-length <int>, -min-pruning\n",
      "                        <int>, -pcr-indel-model <NONE, HOSTILE, AGGRESSIVE,\n",
      "                        CONSERVATIVE>, -standard-min-confidence-threshold-for-\n",
      "                        calling <int>(e.g. --haplotypecaller-options=\"-min-\n",
      "                        pruning 4 -standard-min-confidence-threshold-for-\n",
      "                        calling 30\"). (default: None)\n",
      "  --static-quantized-quals STATIC_QUANTIZED_QUALS\n",
      "                        Use static quantized quality scores to a given number\n",
      "                        of levels. Repeat this option multiple times for\n",
      "                        multiple bins. (default: None)\n",
      "  --gvcf                Generate variant calls in .gvcf Format. (default:\n",
      "                        None)\n",
      "  --batch               Given an input list of BAMs, run the variant calling\n",
      "                        of each BAM using one GPU, and process BAMs in\n",
      "                        parallel based on how many GPUs the system has.\n",
      "                        (default: None)\n",
      "  --disable-read-filter DISABLE_READ_FILTER\n",
      "                        Disable the read filters for BAM entries. Currently,\n",
      "                        the supported read filters that can be disabled are\n",
      "                        MappingQualityAvailableReadFilter,\n",
      "                        MappingQualityReadFilter,\n",
      "                        NotSecondaryAlignmentReadFilter, and\n",
      "                        WellformedReadFilter. (default: None)\n",
      "  --max-alternate-alleles MAX_ALTERNATE_ALLELES\n",
      "                        Maximum number of alternate alleles to genotype.\n",
      "                        (default: None)\n",
      "  -G ANNOTATION_GROUP, --annotation-group ANNOTATION_GROUP\n",
      "                        The groups of annotations to add to the output variant\n",
      "                        calls. Currently supported annotation groups are\n",
      "                        StandardAnnotation, StandardHCAnnotation, and\n",
      "                        AS_StandardAnnotation. (default: None)\n",
      "  -GQB GVCF_GQ_BANDS, --gvcf-gq-bands GVCF_GQ_BANDS\n",
      "                        Exclusive upper bounds for reference confidence GQ\n",
      "                        bands. Must be in the range [1, 100] and specified in\n",
      "                        increasing order. (default: None)\n",
      "  --rna                 Run haplotypecaller optimized for RNA data. (default:\n",
      "                        None)\n",
      "  --dont-use-soft-clipped-bases\n",
      "                        Don't use soft clipped bases for variant calling.\n",
      "                        (default: None)\n",
      "  --read-from-tmp-dir   Read from the temporary files generated by fq2bam.\n",
      "                        (default: None)\n",
      "  --run-partition       Divide the whole genome into multiple partitions and\n",
      "                        run multiple processes at the same time, each on one\n",
      "                        partition. (default: None)\n",
      "  --no-alt-contigs      Get rid of output records for alternate contigs.\n",
      "                        (default: None)\n",
      "  --ploidy PLOIDY       Ploidy assumed for the BAM file. Currently only\n",
      "                        haploid (ploidy 1) and diploid (ploidy 2) are\n",
      "                        supported. (default: 2)\n",
      "  -L INTERVAL, --interval INTERVAL\n",
      "                        Interval within which to call the variants from the\n",
      "                        BAM/CRAM file. All intervals will have a padding of\n",
      "                        100 to get read records, and overlapping intervals\n",
      "                        will be combined. Interval files should be passed\n",
      "                        using the --interval-file option. This option can be\n",
      "                        used multiple times (e.g. \"-L chr1 -L chr2:10000 -L\n",
      "                        chr3:20000+ -L chr4:10000-20000\"). (default: None)\n",
      "  -ip INTERVAL_PADDING, --interval-padding INTERVAL_PADDING\n",
      "                        Amount of padding (in base pairs) to add to each\n",
      "                        interval you are including. (default: None)\n",
      "\n",
      "Run Options:\n",
      "  Options related to resource/verbosity/etc\n",
      "\n",
      "  --logfile LOGFILE     Path to the log file. If not specified, messages will\n",
      "                        only be written to the standard error output.\n",
      "                        (default: None)\n",
      "  --tmp-dir TMP_DIR     Full path to the directory where temporary files will\n",
      "                        be stored. (default: .)\n",
      "  --with-petagene-dir WITH_PETAGENE_DIR\n",
      "                        Full path to the PetaGene installation directory. By\n",
      "                        default, this should have been installed at\n",
      "                        /opt/petagene. Use of this option also requires that\n",
      "                        the PetaLink library has been preloaded by setting the\n",
      "                        LD_PRELOAD environment variable. Optionally set the\n",
      "                        PETASUITE_REFPATH and PGCLOUD_CREDPATH environment\n",
      "                        variables that are used for data and credentials.\n",
      "                        Optionally set the PetaLinkMode environment variable\n",
      "                        that is used to further configure PetaLink, notably\n",
      "                        setting it to \"+write\" to enable outputting compressed\n",
      "                        BAM and .fastq files. (default: None)\n",
      "  --keep-tmp            Do not delete the directory storing temporary files\n",
      "                        after completion. (default: None)\n",
      "  --no-seccomp-override\n",
      "                        Do not override seccomp options for docker. (default:\n",
      "                        None)\n",
      "  --version             View compatible software versions. (default: None)\n",
      "  --num-gpus NUM_GPUS   Number of GPUs to use for a run. (default: 1)\n"
     ]
    }
   ],
   "source": [
    "!pbrun haplotypecaller -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37034272-6a20-471f-b994-a506a749ad52",
   "metadata": {},
   "source": [
    "- vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f076e38d-314a-4395-8cbf-c5e770b3a16f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "/usr/local/parabricks/binaries//bin/htvc /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta /workdir/outputdir/fq2bam_output.bam 1 -o /workdir/outputdir/variants_gatk.vcf -nt 5\n",
      "[PB Info 2023-Nov-29 07:23:17] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:23:17] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-29 07:23:17] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-29 07:23:17] ||                         GPU-GATK4 HaplotypeCaller                        ||\n",
      "[PB Info 2023-Nov-29 07:23:17] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:23:34] 0 /workdir/outputdir/fq2bam_output.bam/workdir/outputdir/variants_gatk.vcf\n",
      "[PB Info 2023-Nov-29 07:23:34] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\tRegions-Processed\tRegions/Minute\n",
      "[PB Info 2023-Nov-29 07:23:44] chr1:92476801\t0.2\t443177\t2659026\n",
      "[PB Info 2023-Nov-29 07:23:54] chr1:237638401\t0.3\t975944\t2927829\n",
      "[PB Info 2023-Nov-29 07:24:04] chr2:126508801\t0.5\t1559676\t3119346\n",
      "[PB Info 2023-Nov-29 07:24:14] chr3:23174401\t0.7\t2188409\t3282609\n",
      "[PB Info 2023-Nov-29 07:24:24] chr3:151358401\t0.8\t2737848\t3285414\n",
      "[PB Info 2023-Nov-29 07:24:34] chr4:103569601\t1.0\t3401470\t3401466\n",
      "[PB Info 2023-Nov-29 07:24:44] chr5:68222401\t1.2\t4053497\t3474426\n",
      "[PB Info 2023-Nov-29 07:24:54] chr6:37171201\t1.3\t4680193\t3510144\n",
      "[PB Info 2023-Nov-29 07:25:04] chr7:2932801\t1.5\t5266582\t3511054\n",
      "[PB Info 2023-Nov-29 07:25:14] chr7:148843201\t1.7\t5897954\t3538772\n",
      "[PB Info 2023-Nov-29 07:25:24] chr8:129686401\t1.8\t6502843\t3547003\n",
      "[PB Info 2023-Nov-29 07:25:34] chr9:136492801\t2.0\t7064255\t3532127\n",
      "[PB Info 2023-Nov-29 07:25:44] chr11:4137601\t2.2\t7683919\t3546424\n",
      "[PB Info 2023-Nov-29 07:25:54] chr12:5467201\t2.3\t8279942\t3548546\n",
      "[PB Info 2023-Nov-29 07:26:04] chr13:24446401\t2.5\t8860634\t3544253\n",
      "[PB Info 2023-Nov-29 07:26:14] chr14:53404801\t2.7\t9423558\t3533833\n",
      "[PB Info 2023-Nov-29 07:26:24] chr16:11966401\t2.8\t10031427\t3540502\n",
      "[PB Info 2023-Nov-29 07:26:34] chr17:57624001\t3.0\t10563024\t3521007\n",
      "[PB Info 2023-Nov-29 07:26:44] chr19:17832001\t3.2\t11071687\t3496320\n",
      "[PB Info 2023-Nov-29 07:26:54] chr21:10468801\t3.3\t11543853\t3463155\n",
      "[PB Info 2023-Nov-29 07:27:04] chrX:80731201\t3.5\t12180582\t3480165\n",
      "[PB Info 2023-Nov-29 07:27:14] Total time taken: 220.641994\n",
      "[PB Info 2023-Nov-29 07:27:14] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:27:14] ||        Program:                         GPU-GATK4 HaplotypeCaller        ||\n",
      "[PB Info 2023-Nov-29 07:27:14] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-29 07:27:14] ||        Start Time:                       Wed Nov 29 07:23:17 2023        ||\n",
      "[PB Info 2023-Nov-29 07:27:14] ||        End Time:                         Wed Nov 29 07:27:14 2023        ||\n",
      "[PB Info 2023-Nov-29 07:27:14] ||        Total Time:                           3 minutes 57 seconds        ||\n",
      "[PB Info 2023-Nov-29 07:27:14] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun haplotypecaller \\\n",
    "      --ref /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "      --in-bam /workdir/outputdir/fq2bam_output.bam \\\n",
    "      --out-variants /workdir/outputdir/variants_gatk.vcf \\\n",
    "      --num-gpus 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5d408cc-ff62-4fd5-85c5-7d6d6cf18481",
   "metadata": {},
   "source": [
    "- gvcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bc65cc13-6ad2-4dcd-b3fd-0b284c2d9414",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "/usr/local/parabricks/binaries//bin/htvc /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta /workdir/outputdir/fq2bam_output.bam 1 -o /workdir/outputdir/variants_gatk.gvcf -nt 5 -g\n",
      "[PB Info 2023-Nov-29 07:36:46] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:36:46] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-29 07:36:46] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-29 07:36:46] ||                         GPU-GATK4 HaplotypeCaller                        ||\n",
      "[PB Info 2023-Nov-29 07:36:46] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:37:03] 0 /workdir/outputdir/fq2bam_output.bam/workdir/outputdir/variants_gatk.gvcf\n",
      "[PB Info 2023-Nov-29 07:37:03] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\tRegions-Processed\tRegions/Minute\n",
      "[PB Info 2023-Nov-29 07:37:13] chr1:28185601\t0.2\t158468\t950808\n",
      "[PB Info 2023-Nov-29 07:37:23] chr1:88473601\t0.3\t427485\t1282455\n",
      "[PB Info 2023-Nov-29 07:37:33] chr1:158520001\t0.5\t642078\t1284156\n",
      "[PB Info 2023-Nov-29 07:37:43] chr1:220267201\t0.7\t914918\t1372377\n",
      "[PB Info 2023-Nov-29 07:37:53] chr2:29716801\t0.8\t1181870\t1418244\n",
      "[PB Info 2023-Nov-29 07:38:03] chr2:91699201\t1.0\t1442022\t1442022\n",
      "[PB Info 2023-Nov-29 07:38:13] chr2:159384001\t1.2\t1737777\t1489521\n",
      "[PB Info 2023-Nov-29 07:38:23] chr2:222292801\t1.3\t2009881\t1507408\n",
      "[PB Info 2023-Nov-29 07:38:33] chr3:40171201\t1.5\t2269385\t1512923\n",
      "[PB Info 2023-Nov-29 07:38:43] chr3:104904001\t1.7\t2547663\t1528597\n",
      "[PB Info 2023-Nov-29 07:38:53] chr3:160425601\t1.8\t2800739\t1527675\n",
      "[PB Info 2023-Nov-29 07:39:03] chr4:26414401\t2.0\t3070990\t1535493\n",
      "[PB Info 2023-Nov-29 07:39:13] chr4:95112001\t2.2\t3369332\t1555074\n",
      "[PB Info 2023-Nov-29 07:39:23] chr4:153700801\t2.3\t3625481\t1553776\n",
      "[PB Info 2023-Nov-29 07:39:33] chr5:37540801\t2.5\t3936514\t1574605\n",
      "[PB Info 2023-Nov-29 07:39:43] chr5:112579201\t2.7\t4235803\t1588426\n",
      "[PB Info 2023-Nov-29 07:39:53] chr5:165115201\t2.8\t4470478\t1577815\n",
      "[PB Info 2023-Nov-29 07:40:03] chr6:38587201\t3.0\t4694906\t1564968\n",
      "[PB Info 2023-Nov-29 07:40:13] chr6:107011201\t3.2\t5000262\t1579030\n",
      "[PB Info 2023-Nov-29 07:40:23] chr6:164740801\t3.3\t5233261\t1569978\n",
      "[PB Info 2023-Nov-29 07:40:33] chr7:63892801\t3.5\t5522697\t1577913\n",
      "[PB Info 2023-Nov-29 07:40:43] chr7:108892801\t3.7\t5752066\t1568745\n",
      "[PB Info 2023-Nov-29 07:40:53] chr8:18086401\t3.8\t6013448\t1568725\n",
      "[PB Info 2023-Nov-29 07:41:03] chr8:68222401\t4.0\t6236633\t1559157\n",
      "[PB Info 2023-Nov-29 07:41:13] chr8:141931201\t4.2\t6545076\t1570818\n",
      "[PB Info 2023-Nov-29 07:41:23] chr9:77726401\t4.3\t6812951\t1572219\n",
      "[PB Info 2023-Nov-29 07:41:33] chr9:131198401\t4.5\t7040320\t1564515\n",
      "[PB Info 2023-Nov-29 07:41:43] chr10:54552001\t4.7\t7329422\t1570590\n",
      "[PB Info 2023-Nov-29 07:41:53] chr10:104836801\t4.8\t7548299\t1561717\n",
      "[PB Info 2023-Nov-29 07:42:03] chr11:35198401\t5.0\t7835819\t1567163\n",
      "[PB Info 2023-Nov-29 07:42:13] chr11:94992001\t5.2\t8081186\t1564100\n",
      "[PB Info 2023-Nov-29 07:42:23] chr12:5467201\t5.3\t8280972\t1552682\n",
      "[PB Info 2023-Nov-29 07:42:33] chr12:57748801\t5.5\t8521543\t1549371\n",
      "[PB Info 2023-Nov-29 07:42:43] chr12:126168001\t5.7\t8794876\t1552036\n",
      "[PB Info 2023-Nov-29 07:42:53] chr13:52075201\t5.8\t8996439\t1542246\n",
      "[PB Info 2023-Nov-29 07:43:03] chr14:29596801\t6.0\t9314673\t1552445\n",
      "[PB Info 2023-Nov-29 07:43:13] chr14:92452801\t6.2\t9584911\t1554309\n",
      "[PB Info 2023-Nov-29 07:43:23] chr15:68529601\t6.3\t9843057\t1554166\n",
      "[PB Info 2023-Nov-29 07:43:33] chr16:23625601\t6.5\t10086648\t1551791\n",
      "[PB Info 2023-Nov-29 07:43:43] chr16:87465601\t6.7\t10317228\t1547583\n",
      "[PB Info 2023-Nov-29 07:43:53] chr17:42120001\t6.8\t10493575\t1535645\n",
      "[PB Info 2023-Nov-29 07:44:03] chr18:196801\t7.0\t10677743\t1525391\n",
      "[PB Info 2023-Nov-29 07:44:13] chr18:65760001\t7.2\t10951709\t1528145\n",
      "[PB Info 2023-Nov-29 07:44:23] chr19:35894401\t7.3\t11159865\t1521799\n",
      "[PB Info 2023-Nov-29 07:44:33] chr20:31060801\t7.5\t11386117\t1518148\n",
      "[PB Info 2023-Nov-29 07:44:43] chr21:18854401\t7.7\t11595642\t1512475\n",
      "[PB Info 2023-Nov-29 07:44:53] chr22:37924801\t7.8\t11803027\t1506769\n",
      "[PB Info 2023-Nov-29 07:45:03] chrX:67718401\t8.0\t12157859\t1519732\n",
      "[PB Info 2023-Nov-29 07:45:13] chrUn_KI270744v1:110401\t8.2\t12572341\t1539470\n",
      "[PB Info 2023-Nov-29 07:45:23] Total time taken: 500.575853\n",
      "[PB Info 2023-Nov-29 07:45:23] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-29 07:45:23] ||        Program:                         GPU-GATK4 HaplotypeCaller        ||\n",
      "[PB Info 2023-Nov-29 07:45:23] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-29 07:45:23] ||        Start Time:                       Wed Nov 29 07:36:46 2023        ||\n",
      "[PB Info 2023-Nov-29 07:45:23] ||        End Time:                         Wed Nov 29 07:45:23 2023        ||\n",
      "[PB Info 2023-Nov-29 07:45:23] ||        Total Time:                           8 minutes 37 seconds        ||\n",
      "[PB Info 2023-Nov-29 07:45:23] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun haplotypecaller \\\n",
    "      --ref /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "      --in-bam /workdir/outputdir/fq2bam_output.bam \\\n",
    "      --gvcf \\\n",
    "      --out-variants /workdir/outputdir/variants_gatk.g.vcf \\\n",
    "      --num-gpus 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9992893-7139-4202-9c8e-cae64b39bae4",
   "metadata": {},
   "source": [
    "\"The key difference between a regular VCF and a GVCF is that the GVCF has records for all sites, whether there is a variant call there or not. The goal is to have every site represented in the file in order to do joint analysis of a cohort in subsequent steps.\" (https://gatk.broadinstitute.org/hc/en-us/articles/360035531812-GVCF-Genomic-Variant-Call-Format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "533cf659-bdae-4e2c-ac82-4700792aa5fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 6.9G\n",
      "drwxr-xr-x. 2 root root  225 Nov 29 07:45 .\n",
      "drwxr-xr-x. 5 root root  216 Nov 29 07:46 ..\n",
      "-rw-r--r--. 1 root root 4.5G Nov 29 07:05 fq2bam_output.bam\n",
      "-rw-r--r--. 1 root root 6.6M Nov 29 07:05 fq2bam_output.bam.bai\n",
      "-rw-r--r--. 1 root root  86K Nov 29 07:05 fq2bam_output_chrs.txt\n",
      "-rw-r--r--. 1 root root 1.6G Nov 29 07:45 variants_gatk.gvcf\n",
      "-rw-r--r--. 1 root root 389K Nov 29 07:45 variants_gatk.gvcf.idx\n",
      "-rw-r--r--. 1 root root  23M Nov 29 07:27 variants_gatk.vcf\n"
     ]
    }
   ],
   "source": [
    "!ls -al -h outputdir/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ca6968e-4835-4138-ae7b-b3294867d3a6",
   "metadata": {},
   "source": [
    "#### DeepVariant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "774b562b-378c-42da-8579-f645f3a7a84e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "usage: pbrun deepvariant <options>\n",
      "Help: pbrun deepvariant -h\n",
      "\n",
      "Run DeepVariant to convert BAM/CRAM to VCF.\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input Output file options:\n",
      "  Options for Input and Output files for this tool:\n",
      "\n",
      "  --ref REF             Path to the reference file. (default: None)\n",
      "  --in-bam IN_BAM       Path to the input BAM/CRAM file for variant calling.\n",
      "                        (default: None)\n",
      "  --interval-file INTERVAL_FILE\n",
      "                        Path to a BED file (.bed) for selective access. This\n",
      "                        option can be used multiple times. (default: None)\n",
      "  --out-variants OUT_VARIANTS\n",
      "                        Path of the vcf/g.vcf/g.vcf.gz file after variant\n",
      "                        calling. (default: None)\n",
      "  --pb-model-file PB_MODEL_FILE\n",
      "                        Path to a non-default parabricks model file for\n",
      "                        deepvariant. (default: None)\n",
      "  --pb-model-dir PB_MODEL_DIR\n",
      "                        Path to a non-default parabricks model dir that\n",
      "                        contains multiple engine files for one model (default:\n",
      "                        None)\n",
      "  --proposed-variants PROPOSED_VARIANTS\n",
      "                        Path of the vcf.gz file, which has proposed variants\n",
      "                        for the make examples stage. (default: None)\n",
      "\n",
      "Tool Options:\n",
      "  Options specific to the tool\n",
      "\n",
      "  --disable-use-window-selector-model\n",
      "                        Change the window selector model from Allele Count\n",
      "                        Linear to Variant Reads. This option will increase the\n",
      "                        accuracy and runtime. (default: None)\n",
      "  --gvcf                Generate variant calls in .gvcf Format. (default:\n",
      "                        None)\n",
      "  --norealign-reads     Do not locally realign reads before calling variants.\n",
      "                        Reads longer than 500 bp are never realigned.\n",
      "                        (default: None)\n",
      "  --sort-by-haplotypes  Reads are sorted by haplotypes (using HP tag).\n",
      "                        (default: None)\n",
      "  --keep-duplicates     Keep reads that are duplicate. (default: None)\n",
      "  --vsc-min-count-snps VSC_MIN_COUNT_SNPS\n",
      "                        SNP alleles occurring at least this many times in the\n",
      "                        AlleleCount will be advanced as candidates. (default:\n",
      "                        2)\n",
      "  --vsc-min-count-indels VSC_MIN_COUNT_INDELS\n",
      "                        Indel alleles occurring at least this many times in\n",
      "                        the AlleleCount will be advanced as candidates.\n",
      "                        (default: 2)\n",
      "  --vsc-min-fraction-snps VSC_MIN_FRACTION_SNPS\n",
      "                        SNP alleles occurring at least this fraction of all\n",
      "                        counts in the AlleleCount will be advanced as\n",
      "                        candidates. (default: 0.12)\n",
      "  --vsc-min-fraction-indels VSC_MIN_FRACTION_INDELS\n",
      "                        Indel alleles occurring at least this fraction of all\n",
      "                        counts in the AlleleCount will be advanced as\n",
      "                        candidates. (default: None)\n",
      "  --min-mapping-quality MIN_MAPPING_QUALITY\n",
      "                        By default, reads with any mapping quality are kept.\n",
      "                        Setting this field to a positive integer i will only\n",
      "                        keep reads that have a MAPQ >= i. Note this only\n",
      "                        applies to aligned reads. (default: 5)\n",
      "  --min-base-quality MIN_BASE_QUALITY\n",
      "                        Minimum base quality. This option enforces a minimum\n",
      "                        base quality score for alternate alleles. Alternate\n",
      "                        alleles will only be considered if all bases in the\n",
      "                        allele have a quality greater than min_base_quality.\n",
      "                        (default: 10)\n",
      "  --mode MODE           Value can be one of [shortread, pacbio, ont]. By\n",
      "                        default, it is shortread. If mode is set to pacbio,\n",
      "                        the following defaults are used: --norealign-reads,\n",
      "                        --alt-aligned-pileup diff_channels, --vsc-min-\n",
      "                        fraction-indels 0.12. If mode is set to ont, the\n",
      "                        following defaults are used: -norealign-reads,\n",
      "                        --variant-caller VCF_CANDIDATE_IMPORTER. (default:\n",
      "                        shortread)\n",
      "  --alt-aligned-pileup ALT_ALIGNED_PILEUP\n",
      "                        Value can be one of [none, diff_channels]. Include\n",
      "                        alignments of reads against each candidate alternate\n",
      "                        allele in the pileup image. (default: None)\n",
      "  --variant-caller VARIANT_CALLER\n",
      "                        Value can be one of [VERY_SENSITIVE_CALLER,\n",
      "                        VCF_CANDIDATE_IMPORTER]. The caller to use to make\n",
      "                        examples. If you use VCF_CANDIDATE_IMPORTER, it\n",
      "                        implies force calling. Default is\n",
      "                        VERY_SENSITIVE_CALLER. (default: None)\n",
      "  --add-hp-channel      Add another channel to represent HP tags per read.\n",
      "                        (default: None)\n",
      "  --parse-sam-aux-fields\n",
      "                        Auxiliary fields of the BAM/CRAM records are parsed.\n",
      "                        If either --sort-by-haplotypes or --add-hp-channel is\n",
      "                        set, then this option must also be set. (default:\n",
      "                        None)\n",
      "  --use-wes-model       If passed, the WES model file will be used. Only used\n",
      "                        in shortread mode. (default: None)\n",
      "  --run-partition       Divide the whole genome into multiple partitions and\n",
      "                        run multiple processes at the same time, each on one\n",
      "                        partition. (default: None)\n",
      "  --include-med-dp      If True, include MED_DP in the output gVCF records.\n",
      "                        (default: None)\n",
      "  --normalize-reads     If True, allele counter left align INDELs for each\n",
      "                        read. (default: None)\n",
      "  --channel-insert-size\n",
      "                        If True, add insert_size channel into pileup image. By\n",
      "                        default, this parameter is true in WGS and WES mode.\n",
      "                        (default: None)\n",
      "  --no-channel-insert-size\n",
      "                        If True, don't add insert_size channel into the pileup\n",
      "                        image. (default: None)\n",
      "  --max-read-size-512   Allow deepvariant to run on reads of size 512bp. The\n",
      "                        default size is 320 bp. (default: None)\n",
      "  --prealign-helper-thread\n",
      "                        Use an extra thread for the pre-align step. This\n",
      "                        parameter is more useful when --max-reads-size-512 is\n",
      "                        set. (default: None)\n",
      "  --max-reads-per-partition MAX_READS_PER_PARTITION\n",
      "                        The maximum number of reads per partition that are\n",
      "                        considered before following processing such as\n",
      "                        sampling and realignment. (default: 1500)\n",
      "  --partition-size PARTITION_SIZE\n",
      "                        The maximum number of basepairs allowed in a region\n",
      "                        before splitting it into multiple smaller subregions.\n",
      "                        (default: 1000)\n",
      "  --track-ref-reads     If True, allele counter keeps track of reads\n",
      "                        supporting ref. By default, allele counter keeps a\n",
      "                        simple count of the number of reads supporting ref.\n",
      "                        (default: None)\n",
      "  --phase-reads         Calculate phases and add HP tag to all reads\n",
      "                        automatically. (default: None)\n",
      "  --dbg-min-base-quality DBG_MIN_BASE_QUALITY\n",
      "                        Minimum base quality in a k-mer sequence to consider.\n",
      "                        (default: 15)\n",
      "  --ws-min-windows-distance WS_MIN_WINDOWS_DISTANCE\n",
      "                        Minimum distance between candidate windows for local\n",
      "                        assembly (default: 80)\n",
      "  --channel-gc-content  If True, add gc_content channel into pileup image\n",
      "                        (default: None)\n",
      "  --channel-hmer-deletion-quality\n",
      "                        If True, add hmer deletion quality channel into pileup\n",
      "                        image (default: None)\n",
      "  --channel-hmer-insertion-quality\n",
      "                        If True, add hmer insertion quality channel into\n",
      "                        pileup image (default: None)\n",
      "  --channel-non-hmer-insertion-quality\n",
      "                        If True, add non-hmer insertion quality channel into\n",
      "                        pileup image (default: None)\n",
      "  --skip-bq-channel     If True, ignore base quality channel. (default: None)\n",
      "  --aux-fields-to-keep AUX_FIELDS_TO_KEEP\n",
      "                        Comma-delimited list of auxiliary BAM fields to keep.\n",
      "                        Values can be [HP, tp, t0] (default: HP)\n",
      "  --vsc-min-fraction-hmer-indels VSC_MIN_FRACTION_HMER_INDELS\n",
      "                        Hmer Indel alleles occurring at least this be advanced\n",
      "                        as candidates. Use this threshold if hmer and non-hmer\n",
      "                        indels should be treated differently (Ultima\n",
      "                        reads)Default will use the same threshold for hmer and\n",
      "                        non-hmer indels, as defined in\n",
      "                        vsc_min_fraction_indels. (default: None)\n",
      "  --vsc-turn-on-non-hmer-ins-proxy-support\n",
      "                        Add read-support from soft-clipped reads and other\n",
      "                        non-hmer insertion alleles,to the most frequent non-\n",
      "                        hmer insertion allele. (default: None)\n",
      "  --consider-strand-bias\n",
      "                        If True, expect SB field in calls and write it to vcf\n",
      "                        (default: None)\n",
      "  --p-error P_ERROR     Basecalling error for reference confidence model.\n",
      "                        (default: 0.001)\n",
      "  --channel-ins-size    If true, add another channel to represent size of\n",
      "                        insertions. (good for flow-based sequencing) (default:\n",
      "                        None)\n",
      "  --max-ins-size MAX_INS_SIZE\n",
      "                        Max insertion size for ins_size_channel, larger\n",
      "                        insertions will look like max (have max intensity)\n",
      "                        (default: 10)\n",
      "  --disable-group-variants\n",
      "                        If using vcf_candidate_importer and multi-allelic\n",
      "                        sites are split across multiple lines in VCF, set to\n",
      "                        True so that variants are not grouped when\n",
      "                        transforming CallVariantsOutput to Variants. (default:\n",
      "                        None)\n",
      "  -L INTERVAL, --interval INTERVAL\n",
      "                        Interval within which to call the variants from the\n",
      "                        BAM/CRAM file. Overlapping intervals will be combined.\n",
      "                        Interval files should be passed using the --interval-\n",
      "                        file option. This option can be used multiple times\n",
      "                        (e.g. \"-L chr1 -L chr2:10000 -L chr3:20000+ -L\n",
      "                        chr4:10000-20000\"). (default: None)\n",
      "\n",
      "Run Options:\n",
      "  Options related to resource/verbosity/etc\n",
      "\n",
      "  --logfile LOGFILE     Path to the log file. If not specified, messages will\n",
      "                        only be written to the standard error output.\n",
      "                        (default: None)\n",
      "  --tmp-dir TMP_DIR     Full path to the directory where temporary files will\n",
      "                        be stored. (default: .)\n",
      "  --with-petagene-dir WITH_PETAGENE_DIR\n",
      "                        Full path to the PetaGene installation directory. By\n",
      "                        default, this should have been installed at\n",
      "                        /opt/petagene. Use of this option also requires that\n",
      "                        the PetaLink library has been preloaded by setting the\n",
      "                        LD_PRELOAD environment variable. Optionally set the\n",
      "                        PETASUITE_REFPATH and PGCLOUD_CREDPATH environment\n",
      "                        variables that are used for data and credentials.\n",
      "                        Optionally set the PetaLinkMode environment variable\n",
      "                        that is used to further configure PetaLink, notably\n",
      "                        setting it to \"+write\" to enable outputting compressed\n",
      "                        BAM and .fastq files. (default: None)\n",
      "  --keep-tmp            Do not delete the directory storing temporary files\n",
      "                        after completion. (default: None)\n",
      "  --no-seccomp-override\n",
      "                        Do not override seccomp options for docker. (default:\n",
      "                        None)\n",
      "  --version             View compatible software versions. (default: None)\n",
      "  --num-gpus NUM_GPUS   Number of GPUs to use for a run. (default: 1)\n"
     ]
    }
   ],
   "source": [
    "!pbrun deepvariant -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d947725a-7a36-41cf-a1b2-feb33fcf8d3b",
   "metadata": {},
   "source": [
    "- vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fedb93a4-8ca8-4afa-9069-938a2052d457",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "Detected 1 CUDA Capable device(s), considering 1 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.2 / 12.0\n",
      "Using model for CUDA Capability Major/Minor version number:    89\n",
      "CUDA_VISIBLE_DEVICES=0 /usr/local/parabricks/binaries//bin/deepvariant /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta /workdir/outputdir/fq2bam_output.bam 1 2 -o 0.vcf  -L \"chr1\" -L \"chr2\" -L \"chr3\" -L \"chr4\" -L \"chr5\" -L \"chr6\" -L \"chr7\" -L \"chrX\" -L \"chr8\" -L \"chr9\" -L \"chr11\" -L \"chr10\" -L \"chr12\" -L \"chr13\" -L \"chr14\" -L \"chr15\" -L \"chr16\" -L \"chr17\" -L \"chr18\" -L \"chr20\" -L \"chr19\" -L \"chrY\" -L \"chr22\" -L \"chr21\" -L \"chrM\" -n 6 --model /usr/local/parabricks/binaries//model/80+/shortread/deepvariant.eng --channel_insert_size --pileup_image_width 221 --max_reads_per_partition 1500 --partition_size 1000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.12 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup none --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10   \n",
      "[PB Info 2023-Nov-30 03:22:57] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:22:57] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-30 03:22:57] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-30 03:22:57] ||                                deepvariant                               ||\n",
      "[PB Info 2023-Nov-30 03:22:57] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:22:57] Starting DeepVariant\n",
      "[PB Info 2023-Nov-30 03:22:57] Running with 1 GPU devices, each with 2 group instances and 6 workers\n",
      "[PB Info 2023-Nov-30 03:23:15] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2023-Nov-30 03:23:21] ProgressMeter -\tchr1:84514000\t0.1\n",
      "[PB Info 2023-Nov-30 03:23:27] ProgressMeter -\tchr1:221012000\t0.2\n",
      "[PB Info 2023-Nov-30 03:23:33] ProgressMeter -\tchr2:98381000\t0.3\n",
      "[PB Info 2023-Nov-30 03:23:39] ProgressMeter -\tchr2:233769000\t0.4\n",
      "[PB Info 2023-Nov-30 03:23:45] ProgressMeter -\tchr3:101383000\t0.5\n",
      "[PB Info 2023-Nov-30 03:23:51] ProgressMeter -\tchr4:35893000\t0.6\n",
      "[PB Info 2023-Nov-30 03:23:57] ProgressMeter -\tchr4:161946000\t0.7\n",
      "[PB Info 2023-Nov-30 03:24:03] ProgressMeter -\tchr5:124932000\t0.8\n",
      "[PB Info 2023-Nov-30 03:24:09] ProgressMeter -\tchr6:78953000\t0.9\n",
      "[PB Info 2023-Nov-30 03:24:15] ProgressMeter -\tchr7:427000\t1.0\n",
      "[PB Info 2023-Nov-30 03:24:21] ProgressMeter -\tchr7:114886000\t1.1\n",
      "[PB Info 2023-Nov-30 03:24:27] ProgressMeter -\tchr8:56103000\t1.2\n",
      "[PB Info 2023-Nov-30 03:24:33] ProgressMeter -\tchr9:601000\t1.3\n",
      "[PB Info 2023-Nov-30 03:24:39] ProgressMeter -\tchr9:133883000\t1.4\n",
      "[PB Info 2023-Nov-30 03:24:45] ProgressMeter -\tchr10:100392000\t1.5\n",
      "[PB Info 2023-Nov-30 03:24:51] ProgressMeter -\tchr11:70842000\t1.6\n",
      "[PB Info 2023-Nov-30 03:24:57] ProgressMeter -\tchr12:34362000\t1.7\n",
      "[PB Info 2023-Nov-30 03:25:03] ProgressMeter -\tchr12:118507000\t1.8\n",
      "[PB Info 2023-Nov-30 03:25:09] ProgressMeter -\tchr14:20508000\t1.9\n",
      "[PB Info 2023-Nov-30 03:25:15] ProgressMeter -\tchr15:18952000\t2.0\n",
      "[PB Info 2023-Nov-30 03:25:21] ProgressMeter -\tchr16:54020000\t2.1\n",
      "[PB Info 2023-Nov-30 03:25:27] ProgressMeter -\tchr17:21727000\t2.2\n",
      "[PB Info 2023-Nov-30 03:25:33] ProgressMeter -\tchr18:8232000\t2.3\n",
      "[PB Info 2023-Nov-30 03:25:39] ProgressMeter -\tchr20:1511000\t2.4\n",
      "[PB Info 2023-Nov-30 03:25:45] ProgressMeter -\tchr22:21339000\t2.5\n",
      "[PB Info 2023-Nov-30 03:25:51] ProgressMeter -\tchrX:55143000\t2.6\n",
      "[PB Info 2023-Nov-30 03:25:57] ProgressMeter -\tchrM:4000\t2.7\n",
      "[PB Info 2023-Nov-30 03:26:04] Deepvariant is finished, total time is 168.718 seconds\n",
      "[PB Info 2023-Nov-30 03:26:04] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        Program:                                       deepvariant        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        Start Time:                       Thu Nov 30 03:22:57 2023        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        End Time:                         Thu Nov 30 03:26:04 2023        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ||        Total Time:                            3 minutes 7 seconds        ||\n",
      "[PB Info 2023-Nov-30 03:26:04] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun deepvariant \\\n",
    "    --ref /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-bam /workdir/outputdir/fq2bam_output.bam \\\n",
    "    --out-variants /workdir/outputdir/variants_dv.vcf \\\n",
    "    --num-streams-per-gpu 2 \\\n",
    "    --run-partition \\\n",
    "    --gpu-num-per-partition 1 \\\n",
    "    --num-gpus 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "232ecee4-d01c-4e22-a39f-0b26202ef336",
   "metadata": {},
   "source": [
    "DeepVariant from Parabricks has the ability to use multiple streams on a GPU. The number of streams that can be used depends on the available resources. The default number of streams is set to two but can be increased up to a maximum of six to get better performance. This is something that has to be experimented with, before getting the optimal number on your system. (https://docs.nvidia.com/clara/parabricks/4.1.0/bestperformance.html#best-performance-for-deepvariant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d459010a-4946-4e70-b144-e637299028ad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "Detected 1 CUDA Capable device(s), considering 1 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.2 / 12.0\n",
      "Using model for CUDA Capability Major/Minor version number:    89\n",
      "CUDA_VISIBLE_DEVICES=0 /usr/local/parabricks/binaries//bin/deepvariant /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta /workdir/outputdir/fq2bam_output.bam 1 4 -o 0.vcf  -L \"chr1\" -L \"chr2\" -L \"chr3\" -L \"chr4\" -L \"chr5\" -L \"chr6\" -L \"chr7\" -L \"chrX\" -L \"chr8\" -L \"chr9\" -L \"chr11\" -L \"chr10\" -L \"chr12\" -L \"chr13\" -L \"chr14\" -L \"chr15\" -L \"chr16\" -L \"chr17\" -L \"chr18\" -L \"chr20\" -L \"chr19\" -L \"chrY\" -L \"chr22\" -L \"chr21\" -L \"chrM\" -n 6 --model /usr/local/parabricks/binaries//model/80+/shortread/deepvariant.eng --channel_insert_size --pileup_image_width 221 --max_reads_per_partition 1500 --partition_size 1000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.12 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup none --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10   \n",
      "[PB Info 2023-Nov-30 03:34:50] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:34:50] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-30 03:34:50] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-30 03:34:50] ||                                deepvariant                               ||\n",
      "[PB Info 2023-Nov-30 03:34:50] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:34:50] Starting DeepVariant\n",
      "[PB Info 2023-Nov-30 03:34:50] Running with 1 GPU devices, each with 4 group instances and 6 workers\n",
      "[PB Info 2023-Nov-30 03:35:10] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2023-Nov-30 03:35:16] ProgressMeter -\t:0\t0.1\n",
      "[PB Info 2023-Nov-30 03:35:22] ProgressMeter -\tchr2:138283000\t0.2\n",
      "[PB Info 2023-Nov-30 03:35:28] ProgressMeter -\tchr3:103721000\t0.3\n",
      "[PB Info 2023-Nov-30 03:35:34] ProgressMeter -\tchr4:144007000\t0.4\n",
      "[PB Info 2023-Nov-30 03:35:40] ProgressMeter -\tchr5:158697000\t0.5\n",
      "[PB Info 2023-Nov-30 03:35:46] ProgressMeter -\tchr7:74992000\t0.6\n",
      "[PB Info 2023-Nov-30 03:35:52] ProgressMeter -\tchr8:113130000\t0.7\n",
      "[PB Info 2023-Nov-30 03:35:58] ProgressMeter -\tchr10:7808000\t0.8\n",
      "[PB Info 2023-Nov-30 03:36:04] ProgressMeter -\tchr11:62775000\t0.9\n",
      "[PB Info 2023-Nov-30 03:36:10] ProgressMeter -\tchr12:78455000\t1.0\n",
      "[PB Info 2023-Nov-30 03:36:16] ProgressMeter -\tchr14:103141000\t1.1\n",
      "[PB Info 2023-Nov-30 03:36:22] ProgressMeter -\tchr17:5978000\t1.2\n",
      "[PB Info 2023-Nov-30 03:36:28] ProgressMeter -\tchr18:53913000\t1.3\n",
      "[PB Info 2023-Nov-30 03:36:34] ProgressMeter -\tchr22:19606000\t1.4\n",
      "[PB Info 2023-Nov-30 03:36:41] Deepvariant is finished, total time is 90.839 seconds\n",
      "[PB Info 2023-Nov-30 03:36:41] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        Program:                                       deepvariant        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        Start Time:                       Thu Nov 30 03:34:50 2023        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        End Time:                         Thu Nov 30 03:36:41 2023        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ||        Total Time:                            1 minute 51 seconds        ||\n",
      "[PB Info 2023-Nov-30 03:36:41] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun deepvariant \\\n",
    "    --ref /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-bam /workdir/outputdir/fq2bam_output.bam \\\n",
    "    --out-variants /workdir/outputdir/variants_dv.vcf \\\n",
    "    --num-streams-per-gpu 4 \\\n",
    "    --run-partition \\\n",
    "    --gpu-num-per-partition 1 \\\n",
    "    --num-gpus 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9f60d6c-985c-4d78-ae85-d735d5c1cc46",
   "metadata": {},
   "source": [
    "- gvcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a85f500f-adbc-43ef-b395-fa4a497e66a1",
   "metadata": {},
   "source": [
    "Using the --run-partition, --proposed-variants, and --gvcf options at the same time will lead to a substantial slowdown."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fc8ae15a-562b-48b3-908c-5352ab87c9b7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please visit https://docs.nvidia.com/clara/#parabricks for detailed documentation\n",
      "\n",
      "Detected 1 CUDA Capable device(s), considering 1 device(s)\n",
      "  CUDA Driver Version / Runtime Version          12.2 / 12.0\n",
      "Using model for CUDA Capability Major/Minor version number:    89\n",
      "/usr/local/parabricks/binaries//bin/deepvariant /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta /workdir/outputdir/fq2bam_output.bam 1 4 -o /workdir/outputdir/variants_dv.g.vcf -n 6 --model /usr/local/parabricks/binaries//model/80+/shortread/deepvariant.eng -g --channel_insert_size --pileup_image_width 221 --max_reads_per_partition 1500 --partition_size 1000 --vsc_min_count_snps 2 --vsc_min_count_indels 2 --vsc_min_fraction_snps 0.12 --min_mapping_quality 5 --min_base_quality 10 --alt_aligned_pileup none --variant_caller VERY_SENSITIVE_CALLER --dbg_min_base_quality 15 --ws_min_windows_distance 80 --aux_fields_to_keep HP --p_error 0.001 --max_ins_size 10\n",
      "[PB Info 2023-Nov-30 03:38:13] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:38:13] ||                 Parabricks accelerated Genomics Pipeline                 ||\n",
      "[PB Info 2023-Nov-30 03:38:13] ||                              Version 4.1.1-1                             ||\n",
      "[PB Info 2023-Nov-30 03:38:13] ||                                deepvariant                               ||\n",
      "[PB Info 2023-Nov-30 03:38:13] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:38:13] Starting DeepVariant\n",
      "[PB Info 2023-Nov-30 03:38:13] Running with 1 GPU devices, each with 4 group instances and 6 workers\n",
      "[PB Info 2023-Nov-30 03:38:32] ProgressMeter -\tCurrent-Locus\tElapsed-Minutes\n",
      "[PB Info 2023-Nov-30 03:38:38] ProgressMeter -\t:0\t0.1\n",
      "[PB Info 2023-Nov-30 03:38:44] ProgressMeter -\tchr2:83197000\t0.2\n",
      "[PB Info 2023-Nov-30 03:38:50] ProgressMeter -\tchr3:11669000\t0.3\n",
      "[PB Info 2023-Nov-30 03:38:56] ProgressMeter -\tchr3:150547000\t0.4\n",
      "[PB Info 2023-Nov-30 03:39:02] ProgressMeter -\tchr4:127242000\t0.5\n",
      "[PB Info 2023-Nov-30 03:39:08] ProgressMeter -\tchr5:140376000\t0.6\n",
      "[PB Info 2023-Nov-30 03:39:14] ProgressMeter -\tchr6:136679000\t0.7\n",
      "[PB Info 2023-Nov-30 03:39:20] ProgressMeter -\tchr7:129151000\t0.8\n",
      "[PB Info 2023-Nov-30 03:39:26] ProgressMeter -\tchr8:139831000\t0.9\n",
      "[PB Info 2023-Nov-30 03:39:32] ProgressMeter -\tchr9:128559000\t1.0\n",
      "[PB Info 2023-Nov-30 03:39:38] ProgressMeter -\tchr11:9010000\t1.1\n",
      "[PB Info 2023-Nov-30 03:39:44] ProgressMeter -\tchr12:43441000\t1.2\n",
      "[PB Info 2023-Nov-30 03:39:50] ProgressMeter -\tchr13:52956000\t1.3\n",
      "[PB Info 2023-Nov-30 03:39:56] ProgressMeter -\tchr15:18419000\t1.4\n",
      "[PB Info 2023-Nov-30 03:40:02] ProgressMeter -\tchr16:22149000\t1.5\n",
      "[PB Info 2023-Nov-30 03:40:08] ProgressMeter -\tchr17:67876000\t1.6\n",
      "[PB Info 2023-Nov-30 03:40:14] ProgressMeter -\tchr19:16984000\t1.7\n",
      "[PB Info 2023-Nov-30 03:40:20] ProgressMeter -\tchr21:6557000\t1.8\n",
      "[PB Info 2023-Nov-30 03:40:26] ProgressMeter -\tchrM:4000\t1.9\n",
      "[PB Info 2023-Nov-30 03:40:33] Deepvariant is finished, total time is 120.940 seconds\n",
      "[PB Info 2023-Nov-30 03:40:33] ------------------------------------------------------------------------------\n",
      "[PB Info 2023-Nov-30 03:40:33] ||        Program:                                       deepvariant        ||\n",
      "[PB Info 2023-Nov-30 03:40:33] ||        Version:                                           4.1.1-1        ||\n",
      "[PB Info 2023-Nov-30 03:40:33] ||        Start Time:                       Thu Nov 30 03:38:13 2023        ||\n",
      "[PB Info 2023-Nov-30 03:40:33] ||        End Time:                         Thu Nov 30 03:40:33 2023        ||\n",
      "[PB Info 2023-Nov-30 03:40:33] ||        Total Time:                           2 minutes 20 seconds        ||\n",
      "[PB Info 2023-Nov-30 03:40:33] ------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!pbrun deepvariant \\\n",
    "    --ref /workdir/sample_data/Ref/Homo_sapiens_assembly38.fasta \\\n",
    "    --in-bam /workdir/outputdir/fq2bam_output.bam \\\n",
    "    --out-variants /workdir/outputdir/variants_dv.g.vcf \\\n",
    "    --num-streams-per-gpu 4 \\\n",
    "    --gvcf \\\n",
    "    --num-gpus 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2df97884-e445-4971-895c-11a7bb1e7e8b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
